File: 03.HeapAndPriorityQueue/剑指Offer40.最小的k个数.java
// 测试平台链接：https://leetcode-cn.com/problems/zui-xiao-de-kge-shu-lcof/
class Solution {
    public int[] getLeastNumbers(int[] arr, int k) {
        if (k <= 0 || arr == null || arr.length == 0) {
            return new int[0];
        }

        Queue<Integer> Q = new PriorityQueue<>((v1, v2) -> v2 - v1);
        for (int x: arr) {
            Q.offer(x);
            while (Q.size() > k) {
                Q.poll();
            }
        }

        int[] ans = new int[k];
        int i = 0;
        for (int x: Q) {
            ans[i++] = x;
        }

        return ans;
    }
}

File: 03.HeapAndPriorityQueue/剑指Offer40.最小的k个数.cpp
// 测试平台: https://leetcode-cn.com/problems/zui-xiao-de-kge-shu-lcof/
class Solution {
public:
    vector<int> getLeastNumbers(vector<int>& arr, int k) {
        if (k <= 0 || arr.empty()) {
            return {};
        }

        if (arr.size() <= k) {
            return arr;
        }

        // c++默认是大堆
        priority_queue<int> Q;

        for (auto x: arr) {
            // 所有元素都要入堆
            Q.push(x);
            // 如果超出k个元素
            // 那么将较大的元素扔掉
            while (Q.size() > k) {
                Q.pop();
            }
        }

        // Q中已经保留了最小的k个数，这里生成返回值
        vector<int> ans;
        while (Q.size() > 0) {
            ans.push_back(Q.top());
            Q.pop();
        }

        return ans;
    }
};
